### PR TITLE
metric: replace QueryDurationHistogram's "general" type to more detail stmt type

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

current Prometheus only display QueryDurationHistogram with internal and general, for general we often need more detail about which kind of SQL become slow.

### What is changed and how it works?

replace QueryDurationHistogram's "general" type to more detail stmt type

why choose to replace general? add new dimensions will make time serial x2 per host, +n per host

corner case:

1. SQL like `insert into tbl select a, b from tbl2` will be recorded as root stmt type ---- `Insert`
2. Multiple SQL `sql1;sql2` will use last stmt's type 

### Check List 

Tests 

 - Exists Test Case

Code changes

- Collection root stmt type in ProcessInfo
- Record metric 

Side effects

 - N/A

Related changes

 - Need to update the documentation
 - Need to be included in the release note

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8819)

### CODE REVIEW

File: executor/adapter.go [hunk 1]
```diff
@@ -44,7 +44,7 @@ import (
 )
 
 type processinfoSetter interface {
-	SetProcessInfo(string, time.Time, byte)
+	SetProcessInfo(string, time.Time, byte, string)
```
Comment (shenli) [inline executor/adapter.go]: Please add a comment for the interface.

File: executor/adapter.go [hunk 2]
```diff
@@ -155,6 +155,10 @@ func (a *ExecStmt) OriginText() string {
 	return a.Text
 }
 
+func (a *ExecStmt) rootStmt() string {
```
Comment (shenli) [inline executor/adapter.go]: Please add a comment for it.

File: server/conn.go [hunk 3]
```diff
@@ -625,7 +625,11 @@ func (cc *clientConn) dispatch(data []byte) error {
 	cc.lastCmd = hack.String(data)
 	token := cc.server.getToken()
 	defer func() {
-		cc.ctx.SetProcessInfo("", t, mysql.ComSleep)
+		stmtType := cc.ctx.ShowProcess().StmtType
+		if stmtType != "" {
+			metrics.StmtDurationHistogram.WithLabelValues(stmtType).Observe(time.Since(t).Seconds())
```
Comment (shenli) [inline server/conn.go]: Could we handle the metrics at the same place as QueryDurationHistogram?

**Comment by imtbkcat** [comment]
LGTM

File: server/conn.go [hunk 4]
```diff
@@ -604,7 +604,12 @@ func (cc *clientConn) addMetrics(cmd byte, startTime time.Time, err error) {
 	} else {
 		metrics.QueryTotalCounter.WithLabelValues(label, "OK").Inc()
 	}
-	metrics.QueryDurationHistogram.WithLabelValues(metrics.LblGeneral).Observe(time.Since(startTime).Seconds())
+	stmtType := cc.ctx.GetSessionVars().StmtCtx.StmtType
+	sqlType := metrics.LblGeneral
+	if stmtType != "" {
+		sqlType = stmtType
+	}
+	metrics.QueryDurationHistogram.WithLabelValues(sqlType).Observe(time.Since(startTime).Seconds())
```
Comment (zimulala) [inline server/conn.go]: Why do `QueryDurationHistogram` and `QueryTotalCounter` use the different `label`？

Comment (lysu) [reply server/conn.go]: I think it's a "historical reasons", old code choose COMMAND as count label and none label for the duration, 

To add duration label, IMO stmtType(like insert/select) is more useful than command(like COM_QUERY).

why don't change QueryTotalCounter is maybe sometime we need that, and we also have another `StmtNodeCounter` to cover "count by stmtType" although has some litte different.

so after this we can see:

- count of command ->  QueryTotalCounter
- count of stmt -> StmtNodeCounter
- duration of query -> QueryDurationHistogram and optional switch to stmtType view

:D

**Pull request merged**